java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -959,53 +964,68 @@ class DatabaseImpl(
         schema: Schema,
         db: SQLiteDatabase,
         counters: Counters? = null
-    ): TypeId = schemaMutex.withLock {
-        schemaTypeMap[schema.hash]?.let {
-            counters?.increment(DatabaseCounters.ENTITY_SCHEMA_CACHE_HIT)
-            return@withLock it
-        }
-        counters?.increment(DatabaseCounters.ENTITY_SCHEMA_CACHE_MISS)
-
-        return db.transaction {
-            counters?.increment(DatabaseCounters.INSERT_ENTITY_TYPE_ID)
+    ): TypeId = db.transaction {
+        var cacheHit = false
+        val result = schemaTypeMap.updateAndGet { currentMap ->
+            currentMap[schema.hash]?.let {
+                cacheHit = true
+                return@updateAndGet currentMap
+            }
+            cacheHit = false
             val content = ContentValues().apply {
                 put("name", schema.hash)
                 put("is_primitive", false)
             }
-            val schemaTypeId = insertOrThrow(TABLE_TYPES, null, content)
+            val id = insertWithOnConflict(
+                TABLE_TYPES,
+                null,
+                content,
+                SQLiteDatabase.CONFLICT_IGNORE
+            )
+            if (id >= 0) {
+                currentMap + (schema.hash to id)
+            } else {
+                cacheHit = true
```

Comment:
from https://developer.android.com/reference/android/database/sqlite/SQLiteDatabase#insertWithOnConflict(java.lang.String,%20java.lang.String,%20android.content.ContentValues,%20int)
it looks like it always returns -1 if you use CONFLICT_IGNORE

Reply:
Oooh, I mis-read that the first time. Thank you!